### PR TITLE
osd: Prevent negative local num_bytes sent to peer for backfill reser…

### PR DESCRIPTION
**Pull request by dzafman**
…vation

Signed-off-by: David Zafman <dzafman@redhat.com>

- [x] References tracker ticket
- [x] ~Updates documentation if necessary~
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dzafman** [comment]
Testing passed:
http://pulpito.ceph.com/dzafman-2019-02-16_15:02:14-rados:thrash-wip-zafman-testing-38344-distro-basic-smithi

File: src/osd/PG.cc [hunk 1]
```diff
@@ -2003,8 +2003,8 @@ void PG::activate(ObjectStore::Transaction& t,
 	pi.last_interval_started = info.last_interval_started;
 	pi.history = info.history;
 	pi.hit_set = info.hit_set;
-        // Save num_bytes for reservation request
-        peer_bytes[peer] = pi.stats.stats.sum.num_bytes;
+        // Save num_bytes for reservation request, can't be negative
+        peer_bytes[peer] = std::max<int64_t>(0, pi.stats.stats.sum.num_bytes);
```
Comment (xiexingguo) [inline src/osd/PG.cc]: So the root cause is that peer's num_bytes is not quite reliable (can be totally wrong)?

Comment (dzafman) [reply src/osd/PG.cc]: @xiexingguo There are a couple of possible explanations of how num_bytes can be off.  I think that if a roll forward or log replay could delete something twice, maybe there would be an undercount.  Or maybe something as simple as a corruption.  The num_bytes for a PG can always be fixed later with a pg repair.

**Comment by dzafman** [comment]
@neha-ojha I'm going to add a fix for a very rare case that could cause an invalid local_num_bytes.

**Comment by dzafman** [comment]
TESTING PASSED, nothing related
dzafman-2019-02-19_16:05:50-rados-wip-zafman-testing-38344-distro-basic-smithi
3614650 DEAD: Infrastructure
3614696 DEAD: ????
3614590 Test failure: test_selftest_run (tasks.mgr.test_module_selftest.TestModuleSelftest)
3614470 OSDs exit 1 without logging anything, no core
3614595 SIGHUPs to processes

**Pull request merged**